flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala

```diff
@@ -0,0 +1,312 @@
+    accumulatorState = getRuntimeContext.getState(accumulatorStateDescriptor)
+
+    val timestampStateDescriptor =
+      new ValueStateDescriptor[JList[Long]]("timestampState", classOf[JList[Long]])
+    timestampState = getRuntimeContext.getState(timestampStateDescriptor)
+
+    val mapStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]](
+        "dataState",
+        classOf[Long],
+        classOf[JList[Row]])
+
+    dataState = getRuntimeContext.getMapState(mapStateDescriptor)
+
+    if (isRowTimeType) {
+      implicit val ord: Ordering[(Long, Long)] = Ordering.by(_._2)
+      timestampPriorityQueue = new mutable.PriorityQueue[(Long, Long)]
+    } else {
+      timestampList = new ArrayList[Long]
+    }
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+
+    if (isRowTimeType) {
+      // triggering timestamp for trigger calculation
+      val triggeringTs = ctx.timestamp + allowedLateness
+
+      val lastTriggeringTs = lastTriggeringTsState.value()
+      // check if the data is expired, if not, save the data and register event time timer
+      if (triggeringTs > lastTriggeringTs&& triggeringTs > ctx.timerService.currentWatermark()) {
+        if (dataState.contains(triggeringTs)) {
+          val data = dataState.get(triggeringTs)
+          data.add(input)
+          dataState.put(triggeringTs, data)
+        } else {
+          val data = new ArrayList[Row]()
+          data.add(input)
+          dataState.put(triggeringTs, data)
+          // register event time timer
+          ctx.timerService().registerEventTimeTimer(triggeringTs)
+        }
+      }
+    } else {
+      // processing timestamp
+      val processingTs = ctx.timerService().currentProcessingTime()
+
+      if (dataState.contains(processingTs)) {
+        val data = dataState.get(processingTs)
+        data.add(input)
+        dataState.put(processingTs, data)
+      } else {
+        val data = new ArrayList[Row]()
+        data.add(input)
+        dataState.put(processingTs, data)
+      }
+      process(processingTs, input, out)
+    }
+  }
+
+  override def onTimer(
+    timestamp: Long,
+    ctx: ProcessFunction[Row, Row]#OnTimerContext,
+    out: Collector[Row]): Unit = {
+    // gets all window data from state for the calculation
+    val inputs: JList[Row] = dataState.get(timestamp)
+    if (null != inputs) {
+      var j: Int = 0
+      while (j < inputs.size) {
+        val input = inputs.get(j)
+        process(timestamp, input, out)
+        j += 1
+      }
+    }
+    lastTriggeringTsState.update(timestamp)
+  }
+
+  /**
+    * Processing the current element
+    *
+    * @param timestamp timestamp of current element
+    * @param input     the element to be calculated
+    * @param out       the collector of the results
+    */
+  def process(timestamp: Long, input: Row, out: Collector[Row]): Unit = {
```

Comment:
To be honest, I don't completely understand your approach.

What about the following design (not considering process time which should be addressed in a separate ProcessFunction, IMO.):
- we have three state objects: 1) the accumulator row, 2) a MapState[Long, List[Row]] for not processed data (`toProcess`), 3) a MapState[Long, List[Row]] for processed data which needs to be retracted (`toRetract`).
- processElement() puts the element in the `toProcess` MapState with the original timestamp and registers a timer for `currentWatermark() + 1`. Hence, we only have a single timer which triggers when the next watermark is reached.
- onTimer() is called for the next watermark. We get an iterator over the `toProcess` MapState. For RocksDB the iterator is sorted on the key. We sort-insert the records from the iterator into a `LinkedList` (since the iterator is sorted for RocksDB this will be simple append. For other state backends it will be more expensive but we can tolerate that, IMO).  We do the same for `toRetract` MapState. So we have two sorted lists for data to accumulate and to retract. We go over both sorted lists and accumulate and retract for each step using the accumulator state. Then we emit the new row and move the emitted row from the `toProcess` MapState to the `toRetract` MapState.

This design has the benefit of using RocksDB to sort. Moreover, we could also put only those fields into the toRetract state that need to be retracted instead of the full row.

What do you think?